workspaces/src/error/impls.rs

```diff
@@ -0,0 +1,199 @@
+        match &self.repr {
+            ErrorRepr::Simple(kind) => kind,
+            ErrorRepr::Message { kind, .. } => kind,
+            ErrorRepr::Custom { kind, .. } => kind,
+            ErrorRepr::Full { kind, .. } => kind,
+        }
+    }
+
+    /// Consumes the `Error`, returning its inner error (if any).
+    ///
+    /// If this [`Error`] was constructed via [`Error::custom`] or [`Error::full`]
+    /// then this function will return [`Some`], otherwise it will return [`None`].
+    pub fn into_inner(self) -> Result<Box<dyn std::error::Error + Send + Sync>, Self> {
+        match self.repr {
+            ErrorRepr::Custom { error, .. } => Ok(error),
+            ErrorRepr::Full { error, .. } => Ok(error),
+            _ => Err(self),
+        }
+    }
+
+    /// Attempt to downgrade the inner error to `E` if any.
+    ///
+    /// Returns `Err(self)` if the downcast is not possible
+    pub fn downcast<E: std::error::Error + Send + Sync + 'static>(self) -> Result<E, Self> {
+        if self.downcast_ref::<E>().is_none() {
+            return Err(self);
+        }
+        // Unwrapping is ok here since we already check above that the downcast will work
+        Ok(*self
+            .into_inner()?
+            .downcast()
+            .expect("failed to unwrap downcast"))
+    }
+
+    /// Returns a reference to the inner error wrapped by this error (if any).
+    pub fn get_ref(&self) -> Option<&(dyn std::error::Error + Send + Sync + 'static)> {
+        match &self.repr {
+            ErrorRepr::Custom { error, .. } => Some(error.as_ref()),
+            ErrorRepr::Full { error, .. } => Some(error.as_ref()),
+            _ => None,
+        }
+    }
+
+    /// Downcast this error object by reference.
+    pub fn downcast_ref<E: std::error::Error + Send + Sync + 'static>(&self) -> Option<&E> {
+        self.get_ref()?.downcast_ref()
+    }
+
+    /// Returns a mutable reference to the inner error wrapped by this error (if any).
+    pub fn get_mut(&mut self) -> Option<&mut (dyn std::error::Error + Send + Sync + 'static)> {
+        match &mut self.repr {
+            ErrorRepr::Custom { error, .. } => Some(error.as_mut()),
+            ErrorRepr::Full { error, .. } => Some(error.as_mut()),
+            _ => None,
+        }
+    }
+
+    /// Returns a mutable reference to the inner error (if any) downcasted to the type provided
+    pub fn downcast_mut<T: std::error::Error + Send + Sync + 'static>(&mut self) -> Option<&mut T> {
+        self.get_mut()?.downcast_mut()
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(error: io::Error) -> Self {
+        Self::custom(ErrorKind::Io, error)
+    }
+}
+
+impl From<base64::DecodeError> for Error {
+    fn from(error: base64::DecodeError) -> Self {
+        Self::custom(ErrorKind::DataConversion, error)
+    }
+}
+
+impl From<serde_json::Error> for Error {
+    fn from(error: serde_json::Error) -> Self {
+        Self::custom(ErrorKind::DataConversion, error)
+    }
+}
```

Comment:
Not a huge fan of these impls: they are public API, but we only want them as impl details. Specific problems with `base64::DecodeError`:

* the user might get an `base64::DecodeError` in a completely different part of their application, and they would be able to convert it to `workspaces::Error`. That is, this impl says that any base64 error is necessary a workspaces error, but that's not the case
* we might migrate of `base64` crate to some other impl of base64. We'd have to change public API for that.